File: Algorithms/Trees/Binary Tree Implementation/validateBST.cpp
#include <bits/stdc++.h>
using namespace std;

bool check(TreeNode* root, long long mn, long long mx){
    if(!root) return true;
    if(root->val <= mn || root->val >= mx) return false;
    
    bool left = check(root->left, mn, root->val);
    bool right = check(root->right, root->val, mx);
    return left && right;
}

bool isValidBST(TreeNode* root) {
    return check(root, (long long)INT_MIN-1, (long long)INT_MAX+1);
}

File: CTCI/Chapter 4 Trees and Graphs/validateBST.py
## https://leetcode.com/problems/validate-binary-search-tree/
class Solution:
    LEFTMAX = -2**31 - 1
    RIGHTMAX = 2**31
    
    def validate(self, root, leftBoundary, rightBoundary):
        if not root:
            return True
        if root.val <= leftBoundary or root.val >= rightBoundary:
            return False
        
        isLeftValid = self.validate(root.left, leftBoundary, root.val)
        isRightValid = self.validate(root.right, root.val, rightBoundary)
        return isLeftValid & isRightValid
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.validate(root, self.LEFTMAX, self.RIGHTMAX)